[Clang] [C++26] Implement P2573R2: `= delete("should have a reason");`

Sirraide opened the pull request:
This implements support for the `= delete("message")` syntax that was only just added to C++26 ([P2573R2](https://isocpp.org/files/papers/P2573R2.html#proposal-scope)). 

The major parts are all done, from what I can tell, except for one thing: we need to figure out where to store the message. We’re out of bits in `FunctionDeclBits`, and we can’t put it in the union that stores the `DefaultedFunctionInfo` because a function may end up having to store both at the same time, so I’m thinking we could store it *in* the `DefaultedFunctionInfo` and rename it to `ExtraFunctionInfo` or something like that.

There are some fairly straight-forward things that are still missing; I’m putting these here so I don’t end up forgetting about any of them (I’ll work on these while we figure out where to put the message):
- [x] Add a release note.
- [x] Add a standards citation and update the BNF for `= delete` if we have that in a comment somewhere, etc.
- [x] Feature test macro.
- [x] Update the C++ conformance page.
- [x] Emit a warning in language modes before C++26 (imo there should be no problem w/ supporting this in all language modes where we also support `= delete`).
- [x] More tests for things like allocation functions, which seem to have separate code paths in some cases, as well as for using functions in other ways than just calling them.
- [x] Also test messages containing unicode characters.
- [x] ~~Not ‘necessary’ but would be nice: AST Matchers & libclang support for getting the message.~~ (should probably be a separate pr)

Sirraide committed:
[Clang] Parse `= delete("message")`

Sirraide committed:
[Clang] Tentative serialisation

Sirraide committed:
[Clang] Sema and tests

Sirraide committed:
[Clang] Add ast printer tests

Sirraide committed:
[NFC] clang-format

Sirraide committed:
[Clang] Add release note

Sirraide committed:
[Clang] Move test to the right directory and fix typo

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -8857,7 +8856,7 @@ def err_nontemporal_builtin_must_be_pointer_intfltptr_or_vector : Error<
   "pointer, or a vector of such types (%0 invalid)">;
 
-def err_deleted_function_use : Error<"attempt to use a deleted function">;
+def err_deleted_function_use : Error<"attempt to use a deleted function%select{|: %1}0">;
```
I think we should make the user-provided message part of a note (and keep the error unchanged)

Sirraide replied:
Yeah, I’ve considered that as well; my reasoning here was that the error message is likely what users of this feature would primarily want us to display, and adding it to the error proper rather than just to the note makes it more obvious that this is the message you should pay attention to, if you will (consider also that, iirc, we’re also displaying `static_assert` messages in the error instead of in a note).

Also, as a side note, we already have three or so different notes for this, so we’d either have to edit all of those or add another one; the former is just as much work as changing the errors, and the latter would just add more clutter, in my opinion at least.

cor3ntin replied:
@AaronBallman @erichkeane for additional opinions

erichkeane replied:
I think it should mirror static assert messages, which is the colon + message in the error.  It is 'important' enough IMO to be considered a part of the first diagnostic.

Sirraide replied:
> Also, as a side note, we already have three or so different notes for this, so we’d either have to edit all of those or add another one; the former is just as much work as changing the errors, and the latter would just add more clutter, in my opinion at least.

Had to change quite a few more diagnostics than I had anticipated, but I’d still argue that while changing the notes would probably be easier, it still makes more sense to display the message in the error proper because that’s what users are most likely to pay attention to (and in some cases, e.g. if the error is displayed in a popup in an IDE, users might not actually be shown any of the notes).

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -18157,7 +18158,7 @@ void Sema::SetDeclDeleted(Decl *Dcl, SourceLocation DelLoc) {
   //  A deleted function is implicitly inline.
   Fn->setImplicitlyInline();
-  Fn->setDeletedAsWritten();
+  Fn->setDeletedWithMessage(Message);
```
I think we should just add the Message as parameter to `setDeletedAsWritten`

Sirraide replied:
I didn’t want to do that because you’d have to write `setDeletedAsWritten(true, Message)` everywhere, but honestly, that’s probably less confusing than having two ‘entirely different’ functions for this now that I think about it...

Sirraide replied:
> I think we should just add the Message as parameter to `setDeletedAsWritten`

I don’t think that would work too well after all now that the message is stored in the `DefaultFunctionInfo`, because `setDeletedAsWritten()` is called in many places, but the `DefaultFunctionInfo` needs to be constructed separately in some. Currently, there is only one place where `setDeletedAsWritten()` is called and then immediately after `setDeletedMessage()`, so I don’t thing it’s worth merging the two.

cor3ntin commented on clang/include/clang/AST/Decl.h:
```diff
@@ -2013,6 +2013,12 @@ class FunctionDecl : public DeclaratorDecl,
+  ///
+  /// FIXME: Figure out where to actually put this; maybe in the
+  /// 'DefaultedInfo' above?
+  StringLiteral *DeletedMessage;
```
Yes, this is my main concern with this PR, it does seem pretty wasteful.
DefaultedFunctionInfo seem like a good approach to explore

Sirraide replied:
Yeah, this here is definitely just a temporary ‘solution’ to get this working well enough to where it’s testable. We definitely don’t want to leave this here...

Sirraide replied:
Just refactored that; see my comment below.

cor3ntin commented on clang/lib/Parse/Parser.cpp:
```diff
@@ -1473,7 +1475,7 @@ Decl *Parser::ParseFunctionDefinition(ParsingDeclarator &D,
 
   if (BodyKind != Sema::FnBodyKind::Other) {
-    Actions.SetFunctionBodyKind(Res, KWLoc, BodyKind);
+    Actions.SetFunctionBodyKind(Res, KWLoc, BodyKind, DeletedMessage);
```
Here we might want just want to do Res->setDeletedMessage(), or something along these lines?

Sirraide replied:
Hmm, true; it might also be useful to just have that already available anyway in case we end up needing it elsewhere.

Sirraide replied:
So after looking into this a bit, I don’t think we want to do that because `Res` is not a `FunctionDecl*` here, it’s a `Decl*`. Specifically, it could also be a template, which means we’d need to first dig out the actual function using `Sema::AdjustDeclIfTemplate`; the same is also true in at least one other place. `SetFunctionBodyKind` calls `SetDeclDeleted`, which already takes care of all of that, so I think it’d make more sense to put that there (however, a `setDeletedMessage()` on `FunctionDecl` is still useful, and I’ll still be using that in other places).

cor3ntin commented on clang/test/SemaCXX/cxx2c-delete-with-message.cpp:
```diff
@@ -0,0 +1,30 @@
+// RUN: %clang_cc1 -std=c++2c -fsyntax-only -verify %s
```
Can you add tests with unicode in the message?

Sirraide replied:
That didn’t even occur to me; thanks for pointing that out.

Sirraide replied:
Done.

cor3ntin reviewed (commented):
Thanks for working on that, I did a quick preliminary review

cor3ntin commented:
The last bullet point of your todo list should be done as a follow up PR. everything else can/should be done here.
I agree that back-porting this feature seems reasonable

Sirraide commented:
> The last bullet point of your todo list should be done as a follow up PR.

Yeah, that was my thinking too more or less; that’s also why I listed it as ‘not necessary’—and I’m also not that familiar w/ libclang or AST matchers.

ChuanqiXu9 commented on clang/lib/Serialization/ASTWriterDecl.cpp:
```diff
@@ -761,6 +761,13 @@ void ASTDeclWriter::VisitFunctionDecl(FunctionDecl *D) {
+  // add this even though it's 0 in the vast majority of cases. We
+  // might really want to consider storing this in the DefaultedFunctionInfo
+  // instead.
+  if (D->isDeletedAsWritten())
+    Record.AddStmt(D->getDeletedMessage());
```
My instinct reaction to this is that we should better use `AddString` and `readString` pairs.

Sirraide replied:
I was going to go with that, but I couldn’t find any examples of that being done for string literals (at least not in this file), so I wasn’t sure whether that’d be the right approach, but it should be fine since it’s just a string literal.

ChuanqiXu9 replied:
I didn't understand. Since the string literal is required to be `unevaluated string`, I feel it may be good to record the string value directly by `StringLiteral::getBytes`.

Sirraide replied:
> I didn't understand. Since the string literal is required to be `unevaluated string`, I feel it may be good to record the string value directly by `StringLiteral::getBytes`.

Hmm, from what I can tell, we don’t usually do that for string literals though. E.g. the string literal of a `FileScopeASMDecl` is also serialised as an expression:
https://github.com/llvm/llvm-project/blob/ff870aeeb7354fd3f681c17e248131e1065ac407/clang/lib/Serialization/ASTWriterDecl.cpp#L1272-L1274

Also, one issue that I see w/ only storing the string data is that we’d lose e.g. the source location of the string literal.

Fznamznon commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -873,9 +873,14 @@ void DeclPrinter::VisitFunctionDecl(FunctionDecl *D) {
+  else if (D->isDeletedAsWritten()) {
     Out << " = delete";
-  else if (D->isExplicitlyDefaulted())
+    if (const auto *M = D->getDeletedMessage()) {
```
Could you please spell out the type?
```suggestion
    if (const StringLiteral *M = D->getDeletedMessage()) {
```

Sirraide replied:
Ah yeah, I have a habit of always writing `auto`, so I forget about this sometimes; will do.

Fznamznon commented on clang/lib/AST/TextNodeDumper.cpp:
```diff
@@ -1936,6 +1936,9 @@ void TextNodeDumper::VisitFunctionDecl(const FunctionDecl *D) {
   if (D->isTrivial())
     OS << " trivial";
 
+  if (const auto *M = D->getDeletedMessage())
```
Here too. 
```suggestion
  if (const StringLiteral *M = D->getDeletedMessage())
```

Sirraide committed:
[NFC] Spell out type

Co-authored-by: Mariya Podchishchaeva <mariya.podchishchaeva@intel.com>

Sirraide committed:
[NFC] Spell out type

Co-authored-by: Mariya Podchishchaeva <mariya.podchishchaeva@intel.com>

Sirraide committed:
[Clang] Store message in DefaultFunctionInfo

Sirraide commented:
So, the message is now stored in the `DefaultFunctionInfo`, which I’ve renamed to `ExtraFunctionInfo`; because setting the deleted message may require allocating a `DefaultFunctionInfo`, I’ve added a separate function for that (`FunctionDecl::setDeletedMessage()`).

Sirraide committed:
[Clang] Add some unicode tests

Sirraide committed:
[NFC] Update delete syntax in comment

Sirraide committed:
[Clang] Add feature test macro

Sirraide committed:
Merge branch 'main' into delete-with-message

Sirraide committed:
[Clang] Update cxx_status

Sirraide committed:
[Clang] Add extension+compatibility diags

cor3ntin commented on clang/include/clang/AST/Decl.h:
```diff
@@ -2005,12 +2019,12 @@ class FunctionDecl : public DeclaratorDecl,
     LazyDeclStmtPtr Body;
     /// Information about a future defaulted function definition.
-    DefaultedFunctionInfo *DefaultedInfo;
+    ExtraFunctionInfo *ExtraInfo;
```
I prefered the old name. `DefaultedOrDeletedInfo` if you really want. `Extra` is pretty vacuous.

erichkeane replied:
Agreed 100%

Sirraide replied:
Yeah, I didn’t just want to keep it as `DefaultedFunctionInfo` because it’s not just for defaulted functions anymore; `ExtraFunctionInfo` is the best I could come up w/ because I’m bad at naming things... I like `DefaultedOrDeletedInfo`, so I’ll change it to that if there aren’t any objections.

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -3210,7 +3210,11 @@ class Sema final {
     Other,
     ///   = default ;
     Default,
+    ///   deleted-function-body
+    ///
+    /// deleted-function-body:
     ///   = delete ;
+    ///   = delete ( unevaluated-string ) ;
```
Can you update the reference above to point to C++26?

cor3ntin commented on clang/lib/AST/Decl.cpp:
```diff
@@ -3092,30 +3092,58 @@ bool FunctionDecl::isVariadic() const {
 
-void FunctionDecl::setDefaultedFunctionInfo(DefaultedFunctionInfo *Info) {
-  assert(!FunctionDeclBits.HasDefaultedFunctionInfo && "already have this");
+void FunctionDecl::setExtraFunctionInfo(ExtraFunctionInfo *Info) {
```
Ditto, not a fan of this rename.

cor3ntin commented on clang/lib/AST/Decl.cpp:
```diff
@@ -3092,30 +3092,58 @@ bool FunctionDecl::isVariadic() const {
+  // allocated for the deleted message, since that would require recreating
+  // this, but setExtraFunctionInfo() disallows overwriting an already existing
+  // ExtraFunctionInfo.
+  assert(HasDeletedMessage && "Explicitly deleting defaulted function?");
```
```suggestion
  assert(HasDeletedMessage && "No space to store a delete message in this ExtraFunctionInfo");
```

cor3ntin commented on clang/lib/AST/Decl.cpp:
```diff
@@ -3092,30 +3092,58 @@ bool FunctionDecl::isVariadic() const {
+    ExtraInfo->setDeletedMessage(Message);
+  else
+    setExtraFunctionInfo(
+        ExtraFunctionInfo::Create(getASTContext(), {}, Message));
```
```suggestion
        ExtraFunctionInfo::Create(getASTContext(), /*Lookups=*/{}, Message));
```

cor3ntin commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -16155,7 +16155,10 @@ Decl *Sema::ActOnFinishFunctionBody(Decl *dcl, Stmt *Body,
     ExitFunctionBodyRAII ExitRAII(*this, isLambdaCallOperator(FD));
     if (FD) {
-      FD->setBody(Body);
+      // Do not overwrite the potentially present ExtraInfo of a deleted
+      // function declaration.
+      if (!FD->isDeletedAsWritten())
+        FD->setBody(Body);
```
When does that happen?

Sirraide replied:
`Parser::ParseFunctionDefinition()` calls `Sema::ActOnFinishFunctionBody()` *after* calling `SetFunctionBodyKind()` to set the body as deleted; with this pr, the latter also sets the deleted message which is now stored in the `DefaultedOrDeletedInfo`, but without this change, `Sema::ActOnFinishFunctionBody()` would end up overwriting it with a `nullptr` function body.

cor3ntin replied:
Thanks. I think this is worth clarifying in the comment.

cor3ntin reviewed (commented):
Do we need to modify `ODRHash::AddFunctionDecl` ? (Probably)
Otherwise, I think it's starting to be in great shape.

erichkeane reviewed (commented):
Tests need to be fleshed out, I'd like to see template specializations + partial template specializations as well as special member functions and operators.

Sirraide committed:
[NFC] Rename ExtraFunctionInfo and update comments

Sirraide committed:
Merge branch 'main' into delete-with-message

Sirraide commented:
Currently looking at constructors etc., and this here seems weird to me: Why are we issuing a diagnostic and then calling `llvm_unreachable()`? Isn’t the diagnostic superfluous here in that case? Shouldn’t this just be `assert(Ovl == OR_Deleted && "...");`?
https://github.com/llvm/llvm-project/blob/0f5f931a9b32208a4894da57ea5c7428ead9df8d/clang/lib/Sema/SemaInit.cpp#L10010-L10019

erichkeane commented:
> Currently looking at constructors etc., and this here seems weird to me: Why are we issuing a diagnostic and then calling `llvm_unreachable()`? Isn’t the diagnostic superfluous here in that case? Shouldn’t this just be `assert(Ovl == OR_Deleted && "...");`?
> 
> https://github.com/llvm/llvm-project/blob/0f5f931a9b32208a4894da57ea5c7428ead9df8d/clang/lib/Sema/SemaInit.cpp#L10010-L10019

That IS super strange.  The unreachable should be able to eliminate the whole branch.  I Think you're right that this is a case where someone wasn't thinking about the subtle differences between `assert(false)` and `unreachable`.  We still should keep the diagnostic though for non-asserts builds, else that might result in awkward behavior in case of a bug.

Sirraide commented:
> else that might result in awkward behavior in case of a bug

Isn’t that true for a lot of if not most situations where we’re using `assert`/`llvm_unreachable`, though?

erichkeane commented:
> > else that might result in awkward behavior in case of a bug
> 
> Isn’t that true for a lot of if not most situations where we’re using `assert`/`llvm_unreachable`, though?

Yeah, probably, but this is a case where the person who wrote this was pretty non-confident, so I lean towards not changing it too much.

AaronBallman commented:
> > else that might result in awkward behavior in case of a bug
> 
> Isn’t that true for a lot of if not most situations where we’re using `assert`/`llvm_unreachable`, though?

We've gone around on this one a few times, the latest discussion of it was https://discourse.llvm.org/t/llvm-unreachable-is-widely-misused/60587

FWIW, I think that use of unreachable is wrong and an assertion would make more sense. It would be worth doing some archeology to see how we got the unreachable in the first place as to whether we should replace it with `assert(false);` (conservative approach) or `assert(Ovl == OR_Deleted)` (slightly more aggressive approach).

Sirraide commented:
> FWIW, I think that use of unreachable is wrong and an assertion would make more sense. It would be worth doing some archeology to see how we got the unreachable in the first place as to whether we should replace it with `assert(false);` (conservative approach) or `assert(Ovl == OR_Deleted)` (slightly more aggressive approach).

Yeah, it just seems to me that, generally, any code that follows this pattern:
```c++
if (condition) {
  // ... (assuming no control flow out of this 'if' here of course)
  llvm_unreachable("message");
}
```

should really just be
```c
assert(!condition && "message");
```

My reasoning is simply this: With assertions enabled, we get a crash either way, and with assertions disabled, the `if` statement is going to be deleted anyway; if we replace `llvm_unreachable` with `assert(false)`, then that’s going to get macroed out with assertions disabled, but that also means we have now reached a program state that was supposed to be impossible (if it wasn’t, and we can actually deal with that just fine, then why was the `llvm_unreachable()` there in the first place?), and sure, it wouldn’t be ideal if we ended up accepting incorrect code (or miscompiling correct code) as a result, but also, either of those is also just straight-up a bug, and I’m not sure how much sense trying to recover from a state that we were never supposed to be in in the first place makes...

But also, this isn’t really specific to this part of the codebase, and it’s also out of scope for this pr; we should probably come back to this at some point, though.

Sirraide committed:
[Clang] Handle remaining diagnostics

Sirraide commented:
> Tests need to be fleshed out, I'd like to see template specializations + partial template specializations as well as special member functions and operators.

Forgot to mention it explicitly, but I’ve added more tests and implemented this for all the other sema diagnostics that had anything to do w/ explicitly deleted functions (there are a few that say something along the lines of ‘function has been implicitly deleted’, but I didn’t really look at those since this is about explicitly deleting functions).

cor3ntin commented:
I wanted to make you aware of this new core issue https://cplusplus.github.io/CWG/issues/2876.html (which i think we should have tests for). Thanks

Sirraide committed:
[Clang] Skip deleted function body on error

Sirraide commented:
> I wanted to make you aware of this new core issue https://cplusplus.github.io/CWG/issues/2876.html (which i think we should have tests for). Thanks

So, as of now, this
```c++
using T = void ();
using U = int;

T a = delete ("hello");
U b = delete ("hello"), c, d = delete ("hello");

struct C {
    T e = delete ("hello");
    U f = delete ("hello");
};
```
gives
```
test.cc:4:7: error: only functions can have deleted definitions
    4 | T a = delete ("hello");
      |       ^
test.cc:5:7: error: only functions can have deleted definitions
    5 | U b = delete ("hello"), c, d = delete ("hello");
      |       ^
test.cc:5:32: error: only functions can have deleted definitions
    5 | U b = delete ("hello"), c, d = delete ("hello");
      |                                ^
test.cc:8:8: error: '= delete' is a function definition and must occur in a standalone declaration
    8 |         T e = delete ("hello");
      |               ^
test.cc:9:8: error: cannot delete expression of type 'const char[6]'
    9 |         U f = delete ("hello");
      |               ^      ~~~~~~~~~
```
which seems reasonable to me—unless there’s something I’m missing here, but it’s ill-formed either way, so so long as we’re diagnosing it, we’re fine, from what I can tell at least.

One thing I did just now is made sure we discard the `("message")` if we encounter `= delete` in a place where it doesn’t belong so we don’t issue two errors instead of one (there was a ‘missing semicolon at end of declaration’ error that imo is just noise in this context).

Sirraide commented:
I’ve also added a parser test for that as well just now.

cor3ntin commented:
Thanks for the new tests. did you see this comment? https://github.com/llvm/llvm-project/pull/86526#pullrequestreview-1976690810

Sirraide commented:
> did you see this comment? [#86526 (review)](https://github.com/llvm/llvm-project/pull/86526#pullrequestreview-1976690810)

Thanks for reminding me; I did see that one but ended up forgetting about it because I was sick for pretty much all of last week; I’ll take a look at it later.

Sirraide committed:
[Clang] Include deleted message in ODR hash

Sirraide commented:
Just added the deleted message to the ODR hash as well. Not too familiar w/ ODR checking though, so I wasn’t sure whether to include it only if `SkipBody` is `false`.

cor3ntin commented on clang/lib/AST/ODRHash.cpp:
```diff
@@ -709,6 +712,11 @@ void ODRHash::AddFunctionDecl(const FunctionDecl *Function,
     return;
   }
 
+  if (DeletedMessage) {
+    ID.AddString(DeletedMessage->getBytes());
+    return;
+  }
+
```
I would move that after `AddBoolean(DeletedMessage);` - it make no sense to separate the delete specifier from its message. (there is a question of whether isExplicitlyDefaulted/isDeletedAsWritten should be after skip body, but that seems  logical as is, and at worse it would be pre-existing)

Sirraide replied:
> (there is a question of whether isExplicitlyDefaulted/isDeletedAsWritten should be after skip body, but that seems logical as is, and at worse it would be pre-existing)

Yeah, that’s true; considering those are already always included it makes sense to always include the message as well if it exists.

cor3ntin commented on clang/lib/Frontend/InitPreprocessor.cpp:
```diff
@@ -750,6 +750,9 @@ static void InitializeCPlusPlusFeatureTestMacros(const LangOptions &LangOpts,
   Builder.defineMacro("__cpp_placeholder_variables", "202306L");
 
+  // C++26 features supported in earlier language modes.
+  Builder.defineMacro("__cpp_deleted_function", "202403L");
```
You should modify LanguageExtensions.rst

cor3ntin commented on clang/include/clang/AST/Decl.h:
```diff
@@ -2484,6 +2498,9 @@ class FunctionDecl : public DeclaratorDecl,
 
   void setDeletedAsWritten(bool D = true) { FunctionDeclBits.IsDeleted = D; }
 
+  /// Only valid if isDeletedAsWritten() returns true.
+  void setDeletedMessage(StringLiteral *Message);
+
```
I really think this should just be an extra parameter on `setDeletedAsWritten` instead.

cor3ntin replied:
Yes, I would prefer (my argument is that there is no scenario in which we would call setDeletedMessage without first calling setDeletedAsWritten, and we always have the information available.

Sirraide replied:
> there is no scenario in which we would call setDeletedMessage without first calling setDeletedAsWritten

That’s true; I was thinking we might want to set the message elsewhere at some point, but I think that I was remembering the situation before I included it it in the `DefaultedOrDeletedInfo`. I’ll move it to `setDeletedAsWritten()` and just give it a default argument.

cor3ntin commented on clang/include/clang/AST/Decl.h:
```diff
@@ -1981,21 +1981,35 @@ class FunctionDecl : public DeclaratorDecl,
+      : llvm::TrailingObjects<DefaultedOrDeletedFunctionInfo, DeclAccessPair,
+                              StringLiteral *> {
     friend TrailingObjects;
     unsigned NumLookups;
+    bool HasDeletedMessage;
```
We could probably combine that into a bitfield to save a byte

Sirraide replied:
I don’t think it matters in this case because `sizeof(DefaultedOrDeletedFunctionInfo)` was already 16 before this, meaning this is just using one more byte of padding, from what I can tell, but I can also make it a bitfield seeing as we probably won’t ever need to store 4 billion lookups here.

Sirraide replied:
Actually, 8 not 16, but either way, making it a bitfield doesn’t seem to do much in this case.

cor3ntin replied:
The alignment is likely to go from 32 to 64, isn't it?
I don't insist (at best we would save a few KB per TU, presumably) , but also it doesn't cost much.

Maybe @AaronBallman would have an opinion

Sirraide replied:
> The alignment is likely to go from 32 to 64, isn't it?

It would normally, but from what I can tell, because it’s inheriting from `TrailingObjects`, the alignment is set to the maximum of the alignment of what the type would normally have and the alignment of each of the trailing objects (by an `alignas` on `TralingObjectsBase`), and since the maximum alignment of the latter is `64`, that ends up being the alignment of the overall struct as well and thus also its size.

Printing out `sizeof(DefaultedOrDeletedFunctionInfo)` shows `8` in both cases (w/ and w/o the bitfield). I also don’t have a strong opinion on this personally; I just didn’t want to make it a bitfield when there’s space available anyway because this way, it’s at least obvious that we have more space left if we need to add more flags later.

cor3ntin reviewed (commented):
Last batch of nitpicks, and then I think I'll be happy enough to approve this patch!

Sirraide committed:
[Docs] Update LanguageExtensions.rst

Sirraide committed:
[Clang] Always add deleted message to ODR hash

Sirraide committed:
[Clang] Set deleted message in setDeletedAsWritten()

cor3ntin reviewed (approved):
Thanks for putting up with my comments and working on this feature.
I'm pretty happy with the state of the PR.

Please give it a day or two in case @erichkeane @AaronBallman @shafik @Fznamznon have further comments

erichkeane commented on clang/lib/AST/Decl.cpp:
```diff
@@ -3092,30 +3092,65 @@ bool FunctionDecl::isVariadic() const {
+      std::max({alignof(DefaultedOrDeletedFunctionInfo),
+                alignof(DeclAccessPair), alignof(StringLiteral *)});
+  size_t Size = totalSizeToAlloc<DeclAccessPair, StringLiteral *>(
+      Lookups.size(), !!DeletedMessage);
```
```suggestion
      Lookups.size(), DeletedMessage != nullptr);
```

erichkeane commented on clang/lib/AST/Decl.cpp:
```diff
@@ -3092,30 +3092,65 @@ bool FunctionDecl::isVariadic() const {
+  DefaultedOrDeletedFunctionInfo *Info =
+      new (Context.Allocate(Size, Alignment)) DefaultedOrDeletedFunctionInfo;
   Info->NumLookups = Lookups.size();
+  Info->HasDeletedMessage = !!DeletedMessage;
```
```suggestion
  Info->HasDeletedMessage = DeletedMessage != nullptr;
```

erichkeane commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -14527,20 +14525,24 @@ Sema::CreateOverloadedUnaryOp(SourceLocation OpLoc, UnaryOperatorKind Opc,
     CandidateSet.NoteCandidates(
         PartialDiagnosticAt(OpLoc, PDiag(diag::err_ovl_deleted_oper)
                                        << UnaryOperator::getOpcodeStr(Opc)
+                                       << !!Msg
```
Same a few times here... this is a 'neat' trick, but it is something that we don't ever do, so it is a touch jarring/takes more thinking than something like this should.

Sirraide replied:
> it is something that we don't ever do

Weird, I recall seeing that in other places in the codebase which is the only reason why I used it; I’d normally just write e.g. `bool(Msg)` etc. instead.

erichkeane replied:
> > it is something that we don't ever do
> 
> Weird, I recall seeing that in other places in the codebase which is the only reason why I used it; I’d normally just write e.g. `bool(Msg)` or `Msg != nullptr` etc. instead.

Yikes!  If you see it, feel free to NFC fix those :)  I personally find that pattern absurdly distracting, and it isn't really necessary most of the time (other than with our operator <<'s), since pointer-to-bool conversions are well defined.  The `!= nullptr` is, IMO, the 'most readable' and clear.

Sirraide replied:
> Weird, I recall seeing that in other places in the codebase

E.g.
https://github.com/llvm/llvm-project/blob/938a73422e0b964eba16f272acdfae1d0281772c/clang/lib/AST/ExprConstant.cpp#L2260-L2262

erichkeane reviewed (approved):
1 set of 'nits' a few places, else LGTM.

Sirraide committed:
[NFC] Replace `!!` with `!= nullptr`

Sirraide committed:
[NFC] clang-format

AaronBallman commented on clang/lib/AST/TextNodeDumper.cpp:
```diff
@@ -1936,6 +1936,9 @@ void TextNodeDumper::VisitFunctionDecl(const FunctionDecl *D) {
   if (D->isTrivial())
     OS << " trivial";
 
+  if (const StringLiteral *M = D->getDeletedMessage())
```
You should do similar in JSONNodeDumper.cpp as well.

AaronBallman commented on clang/lib/Parse/ParseCXXInlineMethods.cpp:
```diff
@@ -70,8 +108,14 @@ NamedDecl *Parser::ParseCXXInlineMethodDef(
+      Actions.SetDeclDeleted(FnD, KWLoc, Message);
       Delete = true;
+      if (Message)
+        Diag(Message->getBeginLoc(), getLangOpts().CPlusPlus26
```
I think should be implemented within `ParseCXXDeletedFunctionMessage()` so anyone calling it gets the diagnostic.

Sirraide replied:
Oh yeah, that makes a lot more sense; I have no idea why I didn’t put it there candidly.

AaronBallman commented on clang/lib/Parse/Parser.cpp:
```diff
@@ -1415,6 +1416,12 @@ Decl *Parser::ParseFunctionDefinition(ParsingDeclarator &D,
           << 1 /* deleted */;
       BodyKind = Sema::FnBodyKind::Delete;
+      DeletedMessage = ParseCXXDeletedFunctionMessage();
+      if (DeletedMessage)
+        Diag(DeletedMessage->getBeginLoc(),
+             getLangOpts().CPlusPlus26 ? diag::warn_cxx23_delete_with_message
+                                       : diag::ext_delete_with_message)
+            << DeletedMessage->getSourceRange();
```
This code can be removed once the diagnostic is hoisted into the helper method.

AaronBallman commented on clang/test/Parser/cxx2c-delete-with-message.cpp:
```diff
@@ -0,0 +1,48 @@
+  S(const S&) = delete("foo"); // pre26-warning {{'= delete' with a message is a C++2c extension}} compat-warning {{'= delete' with a message is incompatible with C++ standards before C++2c}}
+  S(S&&) = delete("foo"); // pre26-warning {{'= delete' with a message is a C++2c extension}} compat-warning {{'= delete' with a message is incompatible with C++ standards before C++2c}}
+  S& operator=(const S&) = delete("foo"); // pre26-warning {{'= delete' with a message is a C++2c extension}} compat-warning {{'= delete' with a message is incompatible with C++ standards before C++2c}}
+  S& operator=(S&&) = delete("foo"); // pre26-warning {{'= delete' with a message is a C++2c extension}} compat-warning {{'= delete' with a message is incompatible with C++ standards before C++2c}}
```
Another test that probably makes sense because I can see people trying it and getting sad:
```
constexpr const char *getMsg() { return "this is a message"; }
void func() = delete(getMsg());
```

AaronBallman reviewed (approved):
In general, this LGTM, thank you! I did spot some minor stuff, but feel free to fix it up when landing.

Sirraide committed:
[Clang] Address Aaron’s comments

Sirraide committed:
Merge branch 'main' into delete-with-message

MitalAshok commented on clang/include/clang/AST/Decl.h:
```diff
@@ -2280,18 +2294,18 @@ class FunctionDecl : public DeclaratorDecl,
-  void setDefaultedFunctionInfo(DefaultedFunctionInfo *Info);
-  DefaultedFunctionInfo *getDefaultedFunctionInfo() const;
+  void setDefaultedOrDeletedInfo(DefaultedOrDeletedFunctionInfo *Info);
+  DefaultedOrDeletedFunctionInfo *getDefalutedOrDeletedInfo() const;
```
@Sirraide There's no reason we can't change this name to fix this typo right?

cor3ntin replied:
That can be fixed as NFC commit. I missed it

Sirraide replied:
> @Sirraide There's no reason we can't change this name to fix this typo right?

Ah, my bad, yeah, that should just be a NFC fix.
